Added Protocol how-to doc (#594)
OpenFreeEnergy/gufe: docs/how-tos/protocol.rst

How to define a new ``Protocol``
================================

The **gufe** :ref:`Protocols <protocol>` system is designed as an *extensible point* of the library,
allowing developers to write their own methods for performing free energy calculations in a form that can take full advantage of the OpenFE ecosystem.

Our recommendation in this how-to is to start simple;
we will call out areas where you can iterate further on making your ``Protocol`` more sophisticated.

.. image:: ../_static/gufe_protocol_diagram.svg
    :alt: The ``gufe`` protocol system.

Overview: What makes a ``Protocol``
-----------------------------------

> A complete ``Protocol`` implementation requires several interconnected components:

Reviewer: ```suggestion
A complete ``Protocol`` implementation requires the following interconnected components:
```

> 1. **Settings class**: A Pydantic model defining configuration parameters

Reviewer: ```suggestion
1. **Settings class**: A `Pydantic <https://docs.pydantic.dev/latest/>`_ model defining configuration parameters
```

Reviewer: a user probably is familiar with pydantic, but I think adding a link on first mention of an external library is good practice.

2. **Protocol class**: The main class that creates computational workflows
3. **ProtocolUnit classes**: Individual computational steps in the workflow

> 4. **ProtocolResult class**: Container for aggregated results from multiple runs

Reviewer: would it make sense to mirror the order below and link the sections?

The ``Protocol`` creates a :ref:`ProtocolDAG <protocoldag>` (directed acyclic graph) of :ref:`ProtocolUnit <protocolunit>` objects that define the computational workflow.
Multiple ``ProtocolDAG`` executions can be aggregated into a single :ref:`ProtocolResult <protocolresult>` to provide statistical estimates.

Step 1: Define your Settings
-----------------------------

First, create a settings class that inherits from :class:`.Settings`.
This defines all the configuration parameters your protocol needs:

.. code-block:: python

    from gufe.settings import Settings
    import pydantic
    from openff.units import unit

    class MyProtocolSettings(Settings):
        # Required: number of independent simulations
        n_repeats: int = 5
        
        # Protocol-specific parameters
        simulation_length: pydantic.Field(
            default=10.0 * unit.nanosecond,
            description="Length of each simulation"
        )
        
        equilibration_length: pydantic.Field(
            default=1.0 * unit.nanosecond, 
            description="Equilibration time before production"

>         )

Reviewer: are `n_repeats`, `simulation_length`, and `equilibration length` required, or is this just an example of the type of extension one might want to do?

Step 2: Define your ProtocolResult
----------------------------------

> Create a result class that knows how to compute estimates and uncertainties from your protocol's outputs:

Reviewer: ```suggestion
Create a result class that inherits from :ref:`ProtocolResult <protocolresult>`and defines how to compute estimates and uncertainties from your protocol's outputs:
```

Reviewer: I think it's helpful to explicitly link here because it shows that both of these methods must be defined.

    from gufe import ProtocolResult
    from openff.units import unit
    import numpy as np

    class MyProtocolResult(ProtocolResult):
        
        def get_estimate(self) -> unit.Quantity:
            """Calculate the free energy estimate from all runs."""
            # Extract the key results from all completed runs
            free_energies = []
            for run_result in self.data["free_energies"]:
                free_energies.append(run_result)
            
            # Return the mean as our best estimate
            mean_dg = np.mean(free_energies) * unit.kilocalorie_per_mole
            return mean_dg
        
        def get_uncertainty(self) -> unit.Quantity:
            """Calculate the uncertainty from all runs."""
            free_energies = self.data["free_energies"]
            if len(free_energies) < 2:
                return 0.0 * unit.kilocalorie_per_mole
            
            std_err = np.std(free_energies) / np.sqrt(len(free_energies))
            return std_err * unit.kilocalorie_per_mole

Step 3: Define your ProtocolUnits
----------------------------------

Create the computational units that will perform the actual work.

> Each unit should inherit from :class:`.ProtocolUnit` and implement a ``_execute`` method.

Reviewer: ```suggestion
Each unit should inherit from :ref:`.ProtocolUnit <protocolunit>` and implement an ``_execute`` method.
```

I think I prefer to link to the extensible points doc, since you can get from extensible points -> python api easily, but not vice-versa.

.. important :: 

   Use ``ctx.shared`` for large objects that need to be passed between units.
   This avoids serialization issues and improves performance by keeping file paths in the return objects instead of the large objects themselves.

.. code-block:: python

    from gufe import ProtocolUnit

    class SetupUnit(ProtocolUnit):
        """Prepare the system for simulation."""
        
        @staticmethod
        def _execute(ctx, *, stateA, stateB, mapping, settings, **inputs):
            """Set up the alchemical system."""
            import pickle
            from pathlib import Path
            
            # ctx provides scratch and shared directories
            # Use ctx.shared to write files that other units will need
            shared_dir = Path(ctx.shared)
            
            # Your setup logic here...
            # - Create alchemical system from stateA/stateB
            # - Apply the atom mapping 
            # - Set up force field parameters
            prepared_system = ...  # Your setup code here
            topology = ...         # Your topology creation
            coordinates = ...      # Your coordinate preparation

>             

Reviewer: maybe i'm misreading, but this is a bit confusing to me. this docstring says to set up force field parameters, but where does that happen?

Author: How all this happens is going to be very dependent on the tools and methods one is using. I think this is about as far as we can go with recommendations here.

                logs.append(sim_result.outputs["log"])
                
                # Load coordinate files if needed for analysis
                window = sim_result.outputs["window"]
                coords_file = sim_result.outputs["final_coordinates_file"]
                with open(coords_file, 'rb') as f:
                    coords = pickle.load(f)
                final_coords[window] = coords
            
            # Write final analysis results to shared directory
            shared_dir = Path(ctx.shared)
            analysis_file = shared_dir / "final_analysis.pkl"
            
            analysis_data = {
                "total_free_energy": total_free_energy,
                "final_coordinates": final_coords,
                "simulation_logs": logs
            }
            
            with open(analysis_file, 'wb') as f:
                pickle.dump(analysis_data, f)
            
            return {
                "total_free_energy": total_free_energy,
                "analysis_file": str(analysis_file),
                "log": "Analysis completed"
            }

Step 4: Implement your Protocol class
-------------------------------------

> Now create your main ``Protocol`` class that ties everything together:

Reviewer: ```suggestion
Now create your main custom ``Protocol`` class that inherits from :ref`Protocol <protocol>` and ties everything together:
```

    # Create protocol with custom settings
    settings = MyProtocolSettings(n_repeats=10, simulation_length=20*unit.nanosecond)
    protocol = MyProtocol(settings)
    
    # Create a ProtocolDAG for specific chemical systems
    dag = protocol.create(
        stateA=chem_system_a,
        stateB=chem_system_b,
        mapping=atom_mapping
    )
    
    # Execute on a scheduler (not shown)
    # dag_result = scheduler.execute(dag)
    
    # Gather multiple results into final estimate
    # final_result = protocol.gather([dag_result1, dag_result2, ...])

Best practices and tips
-----------------------

1. **Start simple**: Begin with a minimal working implementation and add complexity gradually.

2. **Handle errors gracefully**: Use ``try``/``except`` in ``_execute`` methods and return meaningful error information.

3. **Use the context effectively**: The ``ctx`` parameter provides ``scratch`` (temporary, persists over execution of a single ``ProtocolUnit``) and ``shared`` (persists over execution of the ``ProtocolDAG``) directories.
   Use ``ctx.shared`` for large objects that need to pass between units; store file paths in return objects, not the objects themselves.

> 4. **Test thoroughly**: Write unit tests for your ``ProtocolUnit`` classes independently.

Reviewer: ```suggestion
4. **Test thoroughly**: Write unit tests for your ``ProtocolUnit`` classes early in development.
```

> 5. **Document your settings**: Use Pydantic's ``Field`` with descriptions to document what each setting does.

Reviewer: ```suggestion
5. **Document your settings**: Use Pydantic's `Field() function <https://docs.pydantic.dev/latest/concepts/fields/>`_ with descriptions to document what each setting does.
```

How to define a new ``Protocol``
================================

The **gufe** :ref:`Protocol <protocol>` system is designed as an *extensible point* of the library,

> allowing developers to write their own methods for performing free energy calculations in a form that can take full advantage of the OpenFE ecosystem.

Reviewer 2: ```suggestion
allowing developers to write their own methods for performing calculations in a form that can take full advantage of the OpenFE ecosystem.
```
Not all Protocols implement free energy methods - see MD protocols or Boltz-2, maybe this is explained elsewhere though.

2. **Settings class**: A `Pydantic <https://docs.pydantic.dev/latest/>`_ model defining configuration parameters
3. **ProtocolUnit classes**: Individual computational steps in the workflow
4. **ProtocolResult class**: Container for aggregated results from multiple runs

The ``Protocol`` creates a :ref:`ProtocolDAG <protocoldag>` (directed acyclic graph) of :ref:`ProtocolUnit <protocolunit>` objects that define the computational workflow.
Multiple ``ProtocolDAG`` executions can be aggregated into a single :ref:`ProtocolResult <protocolresult>` to provide statistical estimates.

If this is your first time writing a ``Protocol``, we recommend defining these components in the following step order.
You can of course iterate on these components as you write each one, improving the implementation as you go.

Step 1: Define your Settings
-----------------------------

First, create a settings class that inherits from :class:`.Settings`.
This defines all the configuration parameters your protocol needs:

.. code-block:: python

    from gufe.settings import Settings
    import pydantic
    from openff.units import unit

>     class MyProtocolSettings(Settings):

Reviewer 3: Following this it might be good to introduce the idea of grouping like settings into separate classes, like we do in the other protocols, to make it easier for the user to find a setting or change a whole section of settings.

        """Settings for an example Protocol.

        """
        # the number of independent simulations to perform
        n_repeats: int = 5
        

>         simulation_length: pydantic.Field(

Reviewer 3: I think we should add more details on the unit typing system we have here and introduce things like `FloatQuantity["nanosecond"]`, something like for settings which require units you can use FloatQuantity and provide a default unit which the value will be converted to this also ensures that the provided unit is compatible with what you intend.

Reviewer: as discussed in gufe dev call, `FloatQuantity` will no longer be used as we move to vendoring interchange and using pydantic v2.

Author: I've updated the `Settings` example to use the new typing system!

            default=10.0 * unit.nanosecond,
            description="Length of each simulation"
        )
        
        equilibration_length: pydantic.Field(
            default=1.0 * unit.nanosecond, 
            description="Equilibration time before production"
        )

Step 2: Define your ProtocolResult
----------------------------------

> Create a result class that inherits from :ref:`ProtocolResult <protocolresult>` and defines how to compute estimates and uncertainties from your protocol's outputs:

Reviewer 4: Maybe define this more broadly, e.g. the ProtocolResult could also return file paths/... ; e.g. the MD protocol does not return estimates (returns None), but filepaths, could be other things.

Author: Thanks for this! I've incorporated this into the notes within step 2!

.. code-block:: python

    from gufe import ProtocolResult
    from openff.units import unit
    import numpy as np

    class MyProtocolResult(ProtocolResult):
        

>         def get_estimate(self) -> unit.Quantity:

Author: I've added this as an inline comment!

            """Calculate the free energy estimate from all runs."""
            # Extract the key results from all completed runs
            free_energies = []

>             for run_result in self.data["free_energies"]:

Reviewer 3: Why doesn't this already have a unit on it?

            shared_dir = Path(ctx.shared)
            
            # Your setup logic here...
            # - Create alchemical system from stateA/stateB
            # - Apply the atom mapping 
            # - Set up force field parameters
            prepared_system = ...  # Your setup code here
            topology = ...         # Your topology creation
            coordinates = ...      # Your coordinate preparation
            
            # Write large objects to shared directory instead of returning them
            system_file = shared_dir / "system.pkl"
            topology_file = shared_dir / "topology.pkl" 
            coords_file = shared_dir / "initial_coords.pkl"
            
            with open(system_file, 'wb') as f:
                pickle.dump(prepared_system, f)
            with open(topology_file, 'wb') as f:
                pickle.dump(topology, f)
            with open(coords_file, 'wb') as f:
                pickle.dump(coordinates, f)
            
            return {
                "system_file": str(system_file),
                "topology_file": str(topology_file),
                "initial_coordinates_file": str(coords_file),

>                 "log": "System setup completed"

Reviewer 3: See this related issue and the example in SepTop https://github.com/OpenFreeEnergy/gufe/issues/579

            import pickle
            from pathlib import Path
            
            # Load large objects from files written by setup unit
            with open(setup_result.outputs["system_file"], 'rb') as f:
                system = pickle.load(f)
            with open(setup_result.outputs["topology_file"], 'rb') as f:
                topology = pickle.load(f)
            with open(setup_result.outputs["initial_coordinates_file"], 'rb') as f:
                coordinates = pickle.load(f)
            
            # Your simulation logic here...
            # - Run equilibration for `settings.equilibration_length`
            # - Run production simulation for `settings.simulation_length`
            # - Calculate free energy contribution
            dg_contribution = ...  # Your free energy calculation
            final_coords = ...     # Your final coordinates
            
            # Write output files to shared directory
            shared_dir = Path(ctx.shared)
            final_coords_file = shared_dir / f"final_coords_window_{window}.pkl"
            
            with open(final_coords_file, 'wb') as f:
                pickle.dump(final_coords, f)
            

>             return {

Reviewer 4: Potentially add explanation of ProtocolUnitResult?

                "window": window,
                "log": f"Simulation completed for window {window}"
            }

    class AnalysisUnit(ProtocolUnit):
        """Analyze results from all simulations."""
        
        @staticmethod
        def _execute(ctx, *, simulation_results, settings, **inputs):
            """Combine results from all simulation windows."""
            import pickle
            from pathlib import Path
            
            # simulation_results will be a list of ProtocolUnitResult objects
            total_free_energy = 0.0
            logs = []
            final_coords = {}
            
            for sim_result in simulation_results:
                # Extract numerical results directly
                total_free_energy += sim_result.outputs["free_energy_contribution"]
                logs.append(sim_result.outputs["log"])
                
                # Load coordinate files if needed for analysis
                window = sim_result.outputs["window"]
                coords_file = sim_result.outputs["final_coordinates_file"]
                with open(coords_file, 'rb') as f:
                    coords = pickle.load(f)

>                 final_coords[window] = coords

Author: I've reworked this example to use lambda windows instead, and then combining reduced potentials from each window and feeding to MBAR in the `AnalysisUnit`. I think this is originally what Claude was going for, but agree it didn't exactly add up.

            
            # Write final analysis results to shared directory
            shared_dir = Path(ctx.shared)
            analysis_file = shared_dir / "final_analysis.pkl"
            
            analysis_data = {
                "total_free_energy": total_free_energy,
                "final_coordinates": final_coords,
                "simulation_logs": logs
            }
            
            with open(analysis_file, 'wb') as f:
                pickle.dump(analysis_data, f)
            
            return {
                "total_free_energy": total_free_energy,
                "analysis_file": str(analysis_file),
                "log": "Analysis completed"
            }

Step 4: Implement your Protocol class
-------------------------------------

> Now create your main custom ``Protocol`` class that inherits from :ref:`Protocol <protocol>` and ties everything together:

Reviewer 4: Maybe add explicitly what it "ties together" before showing it in the code below.

            """Aggregate results from multiple ProtocolDAG executions."""
            # This method combines results from multiple independent protocol runs
            # into data that the ProtocolResult can use to compute estimates
            
            free_energies = []
            all_logs = []
            
            for dag_result in protocol_dag_results:
                # Find the terminal (final) unit results
                for unit_result in dag_result.terminal_protocol_unit_results:
                    if unit_result.name == "final_analysis":
                        free_energies.append(
                            unit_result.outputs["total_free_energy"]
                        )
                        all_logs.extend(unit_result.outputs["simulation_logs"])
            
            return {
                "free_energies": free_energies,
                "logs": all_logs
            }

Step 5: Add validation (optional)
----------------------------------

You can add custom validation to check that inputs are compatible with your protocol:

.. code-block:: python

    class MyProtocol(Protocol):
        # ... other methods ...
        

>         def _validate(

Reviewer 3: Also point to the default validate method as it might not be appropriate for this protocol or they might impliment redundant validation.

    protocol = MyProtocol(settings)
    
    # Create a ProtocolDAG for specific chemical systems
    dag = protocol.create(
        stateA=chem_system_a,
        stateB=chem_system_b,
        mapping=atom_mapping
    )
    
    # Execute on a scheduler (not shown)
    # dag_result = scheduler.execute(dag)
    
    # Gather multiple results into final estimate
    # final_result = protocol.gather([dag_result1, dag_result2, ...])

Best practices and tips
-----------------------

1. **Start simple**: Begin with a minimal working implementation and add complexity gradually.

2. **Handle errors gracefully**: Use ``try``/``except`` in ``_execute`` methods and return meaningful error information.

3. **Use the context effectively**: The ``ctx`` parameter provides ``scratch`` (temporary, persists over execution of a single ``ProtocolUnit``) and ``shared`` (persists over execution of the ``ProtocolDAG``) directories.
   Use ``ctx.shared`` for large objects that need to pass between units; store file paths in return objects, not the objects themselves.

4. **Test thoroughly**: Write unit tests for your ``ProtocolUnit`` classes early in development.

> 5. **Document your settings**: Use Pydantic's `Field() function <https://docs.pydantic.dev/latest/concepts/fields/>`_ with descriptions to document what each setting does.

Reviewer: autodoc pydantic also handles `Field()` - I think we can have a suggestion, but ultimately just encourage good documentation
